### ISSUE TITLE
./src/OVAL/probes/unix/file.c 32 vs 64 bit 

### ISSUE THREAD

**Issue by jlcharton** [OpenSCAP/openscap#843]
Hi,

Unless I missed something, I believe that the get_size fonction is incorrect when compiling in 64 bit mode. In such a case, without any special preprocessor flag (i.e. without _FILE_OFFSET_BITS or _LARGEFILE_SOURCE), the st_size member of the struct stat is 64 bit but the code of get_size falls in the default case line 248 : return SEXP_number_newu_32_r(sexp, st->st_size).
Am I right?

BTW, are you sure of #elif defined(LARGE_FILE_SOURCE)? Shouldn't it be _LARGEFILE_SOURCE instead?
Moreover, this doesn't work on AIX since AIX flags for LSF are -D_LARGE_FILES and eventually -D_LARGE_FILE_API.

Now the question is how to fix this only with preprocessor #if tests in a fully portable manner?

**Comment by jlcharton** [comment]
Maybe checking for _ILP32 / _LP64 (Oracle C/C++, AIX XLC/C++) or for \_\_ILP32\_\_ / \_\_LP64\_\_ (GCC) may solve the problem?

That is something like this (pseudo code):
if _LP64 or \_\_LP64\_\_ are defined => st_size is 64 bit
else if (_FILE_OFFSET_BITS defined && _FILE_OFFSET_BITS == 64) or _LARGE_FILES is defined => st_size is 64 bit
else if (_FILE_OFFSET_BITS defined && _FILE_OFFSET_BITS == 32) => st_size is 32 bit
else if (_FILE_OFFSET_BITS defined && _FILE_OFFSET_BITS != 32or64 ) => error
else error => "please implement this case!"

Notes:
I think that -D_LARGEFILE_SOURCE alone is not enough to make off_t a 64 bit integer
On Linux combining -m64 and -D_FILE_OFFSET_BITS=32 gives a 64 bit off_t
On Solaris combining -m64 and -D_FILE_OFFSET_BITS=32 gives a preprocessor error
On AIX, _LARGE_FILES just need to be defined (i.e. just -D_LARGE_FILES without =32or64) to turn 64 bit off_t.

**Comment by jlcharton** [comment]
Or maybe we could simply do it at runtime:
```
if (sizeof(st->st_size) == 8)
    return SEXP_number_newu_64_r(sexp, st->st_size);
else if (sizeof(st->st_size) == 4)
    return SEXP_number_newu_32_r(sexp, st->st_size);
else {
    /* What the hell is this? */
    abort();

    /* NOTREACHED */
    return NULL;
}
```

**Comment by jan-cerny** [comment]
It should be `_LARGEFILE_SOURCE`. https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#Feature-Test-Macros

Solution with `sizeof()` looks like a better and more readable solution in my opinion.

**Comment by jlcharton** [comment]
Ok, so let's do it the sizeof() way...

**Comment by jlcharton** [comment]
Fixed by 4d7c745 (PR #846)
